### ISSUE TITLE
Implementing Copy Mechanism/Pointer Network

### ISSUE THREAD

**Issue by anjishnu** [awslabs/sockeye#139]
Hi, I'm interested in implementing pointer networks/copy mechanism in SockEye and would be interested in knowing what would be the best way to implement it in your framework.

**Comment by fhieber** [comment]
Hi, thanks for your interest in Sockeye!
We welcome pull requests that add new features to the framework. Could you expand a little bit on what you are planning to implement (e.g. references to papers describing the method)? This would help to determine whether we can easily integrate this into the codebase.

**Comment by anjishnu** [comment]
I'm trying to implement something along the lines of the following work and would be interested in knowing what steps are needed to implement it:

1) https://arxiv.org/abs/1603.06393
2) https://arxiv.org/pdf/1706.04326.pdf
3) https://nlp.stanford.edu/pubs/see2017get.pdf
4) https://arxiv.org/abs/1506.03134

Basically the idea is to augment the output softmax with the attention scores at each step

It's mostly useful for dialogue and parsing tasks - and to improve NMT vocabulary by backing off to a character level transliteration model. I've been looking at the codebase to see how best to implement this and RecurrentDecoder.decode_sequence seems to be good place to start.

**Comment by fhieber** [comment]
Thanks for the references. It looks like indeed decode_sequence would be a good place to start for training. decode_step is used for inference and you'd need to adapt it independently.
While there are probably some changes to be done throughout the code base (training/inference, adjusting vocab size, losses) I don't immediately see any obvious blockers (e.g. bucketing). 
For training, decode_sequence returns a tensor of logits where the depth would be larger in your case. The increase in depth is directly tied to the source_seq_len, so bucketing should still work.

**Comment by anjishnu** [comment]
Sounds good - I'll first try to get a working implementation on a toy problem, then try to think of how to generalize it - I'm still trying to wrap my head around the decoder right now-

In line 574 of decoder.py
```
        # concatenate along time axis
        # hidden_concat: (batch_size, target_seq_len, rnn_num_hidden)
        hidden_concat = mx.sym.concat(*hidden_all, dim=1, name="%shidden_concat" % self.prefix)
        # hidden_concat: (batch_size * target_seq_len, rnn_num_hidden)
        hidden_concat = mx.sym.reshape(data=hidden_concat, shape=(-1, self.num_hidden))

        # logits: (batch_size * target_seq_len, target_vocab_size)
        logits = mx.sym.FullyConnected(data=hidden_concat, num_hidden=self.config.vocab_size,
                                       weight=self.cls_w, bias=self.cls_b, name=C.LOGITS_NAME)

```

Why are we transforming the tensor to have the dimensions (batch_size * target_seq_len, target_vocab_size)? Is this some kind of matrix operation optimization?

**Comment by fhieber** [comment]
FullyConnected only supports 2d data and we want to do the projection to output vocab size in one go for each time-step (which is much more efficient than running it n times on n separate outputs, hence the concat).
Its also important to note that FullyConnected happily takes 3d input, but silently drops a dimensions in this case.

**Comment by fhieber** [comment]
Just following up on this: with the latest update to MXNet 0.12, FullyConnected now supports a `flatten=False` argument so that these reshapes are now longer necessary.

I am closing this due to inactivity, but feel free to open again.
